[JSC] Bound function should be unwrapped when calling it from C++

Constellation opened the pull request:
#### 1ef2618fca1d075259cb8c84e88299c1f34b2f41
<pre>
[JSC] Bound function should be unwrapped when calling it from C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=252445">https://bugs.webkit.org/show_bug.cgi?id=252445</a>
rdar://105571741

Reviewed by Tadeu Zagallo.

We already unwrapped nested JSBoundFunctions in 260303@main. As a result, one-level unwrapping is enough to extract the target function in most cases.
This patch adds one-level unwrapping in executeCall so that we can skip JSBoundFunction indirection.

Before: CXX -&gt; vmEntryToNative -&gt; CXX JSBoundFunction implementation -&gt; vmEntryToJS -&gt; targetFunction
After: CXX -&gt; vmEntryToJS -&gt; targetFunction

We extract &quot;isBoundFunction&quot; information when calling getCallData / getConstructData. We use this to unwrap in executeCall.

We observed large improvement in microbenchmarks. And we observed 0.3% progression in Speedometer2.1

                                              ToT                     Patched

    call-from-cpp-bound                160.4848+-0.0830     ^     89.3845+-0.1212        ^ definitely 1.7954x faster
    call-from-cpp-bound-with-args      169.0045+-0.1030     ^    107.8479+-0.0877        ^ definitely 1.5671x faster

* JSTests/microbenchmarks/call-from-cpp-bound-with-args.js: Added.
(test):
* JSTests/microbenchmarks/call-from-cpp-bound.js: Added.
(test):
* Source/JavaScriptCore/API/JSCallbackConstructor.cpp:
(JSC::JSCallbackConstructor::getConstructData):
* Source/JavaScriptCore/API/JSCallbackObjectFunctions.h:
(JSC::JSCallbackObject&lt;Parent&gt;::getConstructData):
(JSC::JSCallbackObject&lt;Parent&gt;::getCallData):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeBoundCall):
(JSC::Interpreter::executeCallImpl):
(JSC::Interpreter::executeCall):
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/runtime/CallData.h:
* Source/JavaScriptCore/runtime/InternalFunction.cpp:
(JSC::InternalFunction::getCallData):
(JSC::InternalFunction::getConstructData):
* Source/JavaScriptCore/runtime/JSBoundFunction.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSFunction.cpp:
(JSC::JSFunction::getCallData):
(JSC::JSFunction::getConstructData):
* Source/JavaScriptCore/runtime/ProxyObject.cpp:
(JSC::ProxyObject::getCallData):
(JSC::ProxyObject::getConstructData):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSDollarVM::finishCreation):
* Source/WebCore/bindings/js/JSHTMLAllCollectionCustom.cpp:
(WebCore::JSHTMLAllCollection::getCallData):
* Source/WebCore/bindings/js/JSPluginElementFunctions.cpp:
(WebCore::pluginElementCustomGetCallData):
* Source/WebCore/bridge/objc/objc_runtime.mm:
(JSC::Bindings::ObjcFallbackObjectImp::getCallData):
* Source/WebCore/bridge/runtime_object.cpp:
(JSC::Bindings::RuntimeObject::getCallData):
(JSC::Bindings::RuntimeObject::getConstructData):

Canonical link: <a href="https://commits.webkit.org/260494@main">https://commits.webkit.org/260494@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b41f66f3a8fa547150b1cbcace02261774ea59f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108469 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17565 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41319 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117574 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/117787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112356 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19016 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8843 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100695 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114237 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14252 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97470 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42211 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/97622 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10381 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30459 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/98452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8501 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11134 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/98452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16528 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50055 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106045 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12720 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/26247 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3949 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebCore/bindings/js/JSPluginElementFunctions.cpp:
```diff
@@ -150,6 +150,7 @@ CallData pluginElementCustomGetCallData(JSHTMLElement* element)
     if (instance && instance->supportsInvokeDefaultMethod()) {
         callData.type = CallData::Type::Native;
         callData.native.function = callPlugin;
+        callData.native.isBoundFunction = false;
```
It would be nice to a constructor in `CallData` that does this

Constellation replied:
Unfortunately, this is union... :(

tadeuzagallo reviewed (approved):
r=me

Constellation committed:
[JSC] Bound function should be unwrapped when calling it from C++
https://bugs.webkit.org/show_bug.cgi?id=252445
rdar://105571741

Reviewed by Tadeu Zagallo.

We already unwrapped nested JSBoundFunctions in 260303@main. As a result, one-level unwrapping is enough to extract the target function in most cases.
This patch adds one-level unwrapping in executeCall so that we can skip JSBoundFunction indirection.

Before: CXX -> vmEntryToNative -> CXX JSBoundFunction implementation -> vmEntryToJS -> targetFunction
After: CXX -> vmEntryToJS -> targetFunction

We extract "isBoundFunction" information when calling getCallData / getConstructData. We use this to unwrap in executeCall.

We observed large improvement in microbenchmarks. And we observed 0.3% progression in Speedometer2.1

                                              ToT                     Patched

    call-from-cpp-bound                160.4848+-0.0830     ^     89.3845+-0.1212        ^ definitely 1.7954x faster
    call-from-cpp-bound-with-args      169.0045+-0.1030     ^    107.8479+-0.0877        ^ definitely 1.5671x faster

* JSTests/microbenchmarks/call-from-cpp-bound-with-args.js: Added.
(test):
* JSTests/microbenchmarks/call-from-cpp-bound.js: Added.
(test):
* Source/JavaScriptCore/API/JSCallbackConstructor.cpp:
(JSC::JSCallbackConstructor::getConstructData):
* Source/JavaScriptCore/API/JSCallbackObjectFunctions.h:
(JSC::JSCallbackObject<Parent>::getConstructData):
(JSC::JSCallbackObject<Parent>::getCallData):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeBoundCall):
(JSC::Interpreter::executeCallImpl):
(JSC::Interpreter::executeCall):
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/runtime/CallData.h:
* Source/JavaScriptCore/runtime/InternalFunction.cpp:
(JSC::InternalFunction::getCallData):
(JSC::InternalFunction::getConstructData):
* Source/JavaScriptCore/runtime/JSBoundFunction.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSFunction.cpp:
(JSC::JSFunction::getCallData):
(JSC::JSFunction::getConstructData):
* Source/JavaScriptCore/runtime/ProxyObject.cpp:
(JSC::ProxyObject::getCallData):
(JSC::ProxyObject::getConstructData):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSDollarVM::finishCreation):
* Source/WebCore/bindings/js/JSHTMLAllCollectionCustom.cpp:
(WebCore::JSHTMLAllCollection::getCallData):
* Source/WebCore/bindings/js/JSPluginElementFunctions.cpp:
(WebCore::pluginElementCustomGetCallData):
* Source/WebCore/bridge/objc/objc_runtime.mm:
(JSC::Bindings::ObjcFallbackObjectImp::getCallData):
* Source/WebCore/bridge/runtime_object.cpp:
(JSC::Bindings::RuntimeObject::getCallData):
(JSC::Bindings::RuntimeObject::getConstructData):

Canonical link: https://commits.webkit.org/260494@main
